lib/puppet/util/autoload.rb

```diff
@@ -67,6 +69,12 @@ def load_file(name, env)
         message = "Could not autoload #{name}: #{detail}"
         Puppet.log_exception(detail, message)
         raise Puppet::Error, message, detail.backtrace
+      ensure
+        # Restore original LOAD_PATH
+        $LOAD_PATH.each do |p|
+          $LOAD_PATH.delete(p) unless load_path.include? p
+        end
+        $LOAD_PATH.uniq!
```

Comment:
The above restore will have performance impact. It should be possible to do this in a more light way fashion. The above has two linear array searches for each entry on the path + the extra uniq set operation at the end.